src/tools/pkcs15-init.c

```diff
@@ -1598,68 +1726,73 @@ static int
 do_generate_key(struct sc_profile *profile, const char *spec)
 {
 	struct sc_pkcs15init_keygen_args keygen_args;
-	unsigned int keybits = 1024;
-	int		r;
+	unsigned int keybits = 0;
+	int r, algorithm = -1;
 
 	memset(&keygen_args, 0, sizeof(keygen_args));
 	keygen_args.pubkey_label = opt_pubkey_label;
 
-	if ((r = init_keyargs(&keygen_args.prkey_args)) < 0)
+	if ((r = init_prkeyargs(&keygen_args.prkey_args)) < 0)
 		return r;
+
+	algorithm = parse_alg_spec(alg_types_asym, spec, &keybits, &keygen_args.prkey_args.key);
+	if (algorithm < 0) {
+		util_error("Invalid symmetric key spec: \"%s\"", spec);
+		return SC_ERROR_INVALID_ARGUMENTS;
+	}
+
+	keygen_args.prkey_args.key.algorithm = algorithm;
 	keygen_args.prkey_args.access_flags |=
 		  SC_PKCS15_PRKEY_ACCESS_SENSITIVE
 		| SC_PKCS15_PRKEY_ACCESS_ALWAYSSENSITIVE
 		| SC_PKCS15_PRKEY_ACCESS_NEVEREXTRACTABLE
 		| SC_PKCS15_PRKEY_ACCESS_LOCAL;
 
-	/* Parse the key spec given on the command line */
-	if (!strncasecmp(spec, "rsa", 3)) {
-		keygen_args.prkey_args.key.algorithm = SC_ALGORITHM_RSA;
-		spec += 3;
-	} else if (!strncasecmp(spec, "dsa", 3)) {
-		keygen_args.prkey_args.key.algorithm = SC_ALGORITHM_DSA;
-		spec += 3;
-	} else if (!strncasecmp(spec, "gost2001", strlen("gost2001"))) {
-		keygen_args.prkey_args.key.algorithm = SC_ALGORITHM_GOSTR3410;
-		keybits = SC_PKCS15_GOSTR3410_KEYSIZE;
+	switch (algorithm) {
+	case SC_ALGORITHM_GOSTR3410:
 		/* FIXME: now only SC_PKCS15_PARAMSET_GOSTR3410_A */
 		keygen_args.prkey_args.params.gost.gostr3410 = SC_PKCS15_PARAMSET_GOSTR3410_A;
-		spec += strlen("gost2001");
-	} else if (!strncasecmp(spec, "ec", 2)) {
-		keygen_args.prkey_args.key.algorithm = SC_ALGORITHM_EC;
-		spec += 2;
-	} else {
-		util_error("Unknown algorithm \"%s\"", spec);
-		return SC_ERROR_INVALID_ARGUMENTS;
+		break;
 	}
 
-	if (*spec == '/' || *spec == '-' || *spec == ':')
-		spec++;
+	r = sc_lock(p15card->card);
+	if (r == 0)
+		r = sc_pkcs15init_generate_key(p15card, profile, &keygen_args, keybits, NULL);
+	sc_unlock(p15card->card);
+	return r;
+}
 
-	if (*spec)   {
-		if (isalpha(*spec) && keygen_args.prkey_args.key.algorithm == SC_ALGORITHM_EC)   {
-			keygen_args.prkey_args.key.u.ec.params.named_curve = strdup(spec);
-			keybits = 0;
-		}
-		else {
-			char	*end;
+#if 0
```

Comment:
Implemented the high level logic of requesting symmetric key generation. No card driver supports it yet, though.